camelot/handlers.py

```diff
@@ -171,14 +179,54 @@ def parse(
             layout_kwargs = {}
 
         tables = []
+        parser = Lattice(**kwargs) if flavor == "lattice" else Stream(**kwargs)
         with TemporaryDirectory() as tempdir:
-            for p in self.pages:
-                self._save_page(self.filepath, p, tempdir)
-            pages = [os.path.join(tempdir, f"page-{p}.pdf") for p in self.pages]
-            parser = Lattice(**kwargs) if flavor == "lattice" else Stream(**kwargs)
-            for p in pages:
-                t = parser.extract_tables(
-                    p, suppress_stdout=suppress_stdout, layout_kwargs=layout_kwargs
-                )
-                tables.extend(t)
+            cpu_count = mp.cpu_count()
+            # Using multiprocessing only when cpu_count > 1 to prevent a stallness issue
+            # when cpu_count is 1
+            if parallel and len(self.pages) > 1 and cpu_count > 1:
+                with mp.get_context("spawn").Pool(processes=cpu_count) as pool:
+                    jobs = [
+                        pool.apply_async(
+                            self._parse_page,
+                            (p, tempdir, parser, suppress_stdout, layout_kwargs)
+                        ) for p in self.pages
```

Comment:
This list comprehension makes the code less readable. 

Also I would opt to move the code inside a `parse_parallel` method.

Reply:
I moved it to a simple loop.

OTOH I would keep the parallel and sequential processing close to each other to be clear at a glance that the arguments are the same:
```python
j = pool.apply_async(self._parse_page,(p, tempdir, parser, suppress_stdout, layout_kwargs))
```
and
```python
t = self._parse_page(p, tempdir, parser, suppress_stdout, layout_kwargs)
```

Reply:
Good thinking! Thanks for your clarification.